server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -788,34 +778,46 @@ void onShardDone() {
                         relevantShardIndices.set(localIndex);
                     }
                 }
-                final Object[] results = new Object[queryPhaseResultConsumer.getNumShards()];
-                for (int i = 0; i < results.length; i++) {
-                    var result = queryPhaseResultConsumer.results.get(i);
-                    if (result == null) {
-                        results[i] = failures.get(i);
-                    } else {
-                        // free context id and remove it from the result right away in case we don't need it anymore
-                        if (result instanceof QuerySearchResult q
-                            && q.getContextId() != null
-                            && relevantShardIndices.get(q.getShardIndex()) == false
-                            && q.hasSuggestHits() == false
-                            && q.getRankShardResult() == null
-                            && searchRequest.searchRequest.scroll() == null
-                            && isPartOfPIT(searchRequest.searchRequest, q.getContextId()) == false) {
-                            if (dependencies.searchService.freeReaderContext(q.getContextId())) {
-                                q.clearContextId();
+                final int resultCount = queryPhaseResultConsumer.getNumShards();
+                out = dependencies.transportService.newNetworkBytesStream();
+                out.setTransportVersion(channel.getVersion());
+                try {
+                    out.writeVInt(resultCount);
```

Comment:
Unfortunately we need to duplicate the `writeTo()` on the message for the time being because of CCS proxy request BwC concerns urgh :) but I think it's not too bad and we can dry this up in a follow-up. This change is about networking, not search really so I didn't want to mess with that here.